### PR TITLE
Add request processing HPA into the queue after processing is finished.

### LINKED ISSUE #72372: HorizontalPodAutoscalers are not processed every
**What happened**:
Resync period for HorizontalPodAutoscaler controller is set to 15 seconds in order to update HPAs every 15 seconds. Sometimes updates are skipped and period between two updates is 30 seconds.

**What you expected to happen**:
HPA should be updated every 15 seconds.

**How to reproduce it (as minimally and precisely as possible)**:
Add debugs to HPA controller showing how often HPAs are processed. Create HPA.
**Anything else we need to know?**:
HPA controller uses rate RateLimitingQueue:
https://github.com/kubernetes/kubernetes/blob/bb7973a34c209f74af398354fb0247598aa20766/pkg/controller/podautoscaler/horizontal.go#L120
New requests are always added with 15 seconds delay:
https://github.com/kubernetes/kubernetes/blob/bb7973a34c209f74af398354fb0247598aa20766/pkg/controller/podautoscaler/horizontal.go#L183
If request for  the given HPA is in the queue then newly added request is skipped. Requests are added and removed from the queue at the same moment so there is a race condition. See following scenario:
00:00:00 - Resync adds request to queue to be executed at 00:00:15
00:00:15 - Resync adds next request to queue to be executed at 00:00:30 but it is skipped because previous request is still in the queue
00:00:15 - Request added at 00:00:00 is executed
00:00:30 - Resync adds next request to queue to be executed at 00:00:45
00:00:45 - Request added at 00:00:30 is executed
In above scenario HPA wasn't updated between 00:00:15 and 00:00:45.

**Environment**:
- Kubernetes version (use `kubectl version`): HEAD
- Cloud provider or hardware configuration: GKE

/kind bug

### PR DESCRIPTION
**Pull request by krzysztof-jastrzebski**
This fixes a bug whth skipping request inserted by resync because previous one hasn't processed yet.

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR fixes a bug with skipping request inserted by resync because previous one hasn't processed yet.
**Which issue(s) this PR fixes**:

Fixes #72372

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Fixes a bug in HPA controller so HPAs are always updated every resyncPeriod (15 seconds).
```

### CODE REVIEW

File: pkg/controller/podautoscaler/horizontal.go
Comment (krzysztof-jastrzebski) [reply pkg/controller/podautoscaler/horizontal.go]: done

**Comment by mwielgus** [comment]
How was this PR tested?

**Comment by krzysztof-jastrzebski** [comment]
I've created K8s build with debugs showing when HPAs are processed, created cluster and HPA. I observed how often HPA is updated. Next I removed HPA just after HPA was processed and verified that HPA is not processed anymore.

**Comment by krzysztof-jastrzebski** [comment]
HPA controller uses rate limiter which always returns ResyncPeriod (currently 15 seconds).
Rate Limiter:
https://github.com/kubernetes/kubernetes/blob/16fb54ef7c10e3865ef5f0af69b2479a084b85c5/pkg/controller/podautoscaler/rate_limiters.go#L38
Creation:
https://github.com/kubernetes/kubernetes/blob/a6ba2ebf85d8673f78220c089aba1290437de2be/pkg/controller/podautoscaler/horizontal.go#L120
In that case every request inserted to queue waits 15 seconds in the queue. If we insert one request then it stays in queue for 15 seconds and all request inserted during 15 seconds are skipped as they would be executed after request which is in queue. After 15 seconds next request can be inserted and it also stays in queue for 15 seconds etc.
All redundant request are skipped, HPA is updated at most every 15 seconds.

I see this is not the best way of using RateLimitingQueue but I would like to do small fix so I can backport it to 1.12. If you think we should use  queue in other way then I'm happy to fix it next PR. Do you have any suggestion how to do that?

File: pkg/controller/podautoscaler/horizontal.go
Comment (krzysztof-jastrzebski) [reply pkg/controller/podautoscaler/horizontal.go]: yes, see comment below

**Comment by liggitt** [comment]
This is a strange use of the rate limiting queue trying to achieve tight guarantees about retry interval.

I'm concerned about doubling the number of queue insertions for systems with large numbers of HPAs. Attempting to enqueue all HPAs in the system every 15 seconds is already very aggressive... this change would attempt to enqueue them all twice every 15 seconds.

Until the structure of this controller can be revisited, can we do something simpler and less likely to impact performance like one of the following:
* set up the HPA `AddEventHandlerWithResyncPeriod` with `resyncPeriod+time.Second`
* set up `NewDefaultHPARateLimiter(resyncPeriod-time.Second)`

Either of those approaches would give `delayingType#waitingLoop` a second to move ready HPA items from `waitingForQueue` into the queue. If that takes longer than a second (for example, under load), HPA processing could still skip a resync interval, but that actually seems like a better outcome to me than doubling queue attempts in pursuit of exact retry intervals.

**Comment by krzysztof-jastrzebski** [comment]
I was thinking about similar solution. The problem with it is why to add/subtract 1 second. resyncPeriod is defined by a flag. Default is 15 seconds but someone can use 1 second resuncPeriod. If we added 1 second then HPAs would be refreshed every 2 seconds instead of 1, we would also process HPA on every change of the objects (rate limiter set to 0) which can be in some cases very often. 
I don't think that adding HPA to queue after processing it is a problem as current implementation of HPA(it is single-threaded) can process less than 10 HPAs per second. Adding 10 request every second to local queue shouldn't be a big deal.

**Comment by liggitt** [comment]
> Default is 15 seconds but someone can use 1 second resyncPeriod. If we added 1 second then HPAs would be refreshed every 2 seconds instead of 1, we would also process HPA on every change of the objects (rate limiter set to 0) which can be in some cases very often.

Setting a 1 second resyncPeriod isn't likely to work well in general, but I agree we wouldn't want to drop that to 0. Setting to a percentage or capping at a minimum value would prevent that.

Unconditional requeuing is confusing to anyone who is familiar with all the other controllers. I'd prefer the smallest change we can make that doesn't take this controller even further away from standard usage of the queue components, so I'm not in favor of the current PR, but will defer to @kubernetes/sig-autoscaling-pr-reviews 

/hold cancel

File: pkg/controller/podautoscaler/horizontal.go
Comment (DirectXMan12) [inline pkg/controller/podautoscaler/horizontal.go]: you should still have a comment as to why we're not forgetting the key, since this works significantly different from normal controllers.

**Review by DirectXMan12 — CHANGES_REQUESTED** [review]
can you expand a bit on why this occurs in the comment?  Use your example from above

**Comment by DirectXMan12** [comment]
please also fix the PR description to remove the unused flags.

**Comment by DirectXMan12** [comment]
I'm inclined to go with @liggitt's suggestion short-term, and think up a better long-term fix (e.g. don't watch HPA, and instead just reconcile every 15 seconds, staggered, or something).

**Comment by krzysztof-jastrzebski** [comment]
@mwielgus done

**Pull request merged**
